spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/adapter/AbstractAdaptableMessageListener.java

```diff
@@ -414,15 +444,33 @@ protected void sendResponse(Channel channel, Address replyTo, Message messageIn)
 		try {
 			this.logger.debug("Publishing response to exchange = [" + replyTo.getExchangeName() + "], routingKey = ["
 					+ replyTo.getRoutingKey() + "]");
-			channel.basicPublish(replyTo.getExchangeName(), replyTo.getRoutingKey(), this.mandatoryPublish,
-					this.messagePropertiesConverter.fromMessageProperties(message.getMessageProperties(), this.encoding),
-					message.getBody());
+			if (this.retryTemplate == null) {
+				doPublish(channel, replyTo, message);
+			}
+			else {
+				final Message messageToSend = message;
+				final AtomicBoolean first = new AtomicBoolean(true);
+				this.retryTemplate.execute(ctx -> {
+					if (this.recoveryCallback != null && first.getAndSet(false)) {
+						ctx.setAttribute(SendRetryContextAccessor.MESSAGE, messageToSend);
+						ctx.setAttribute(SendRetryContextAccessor.ADDRESS, replyTo);
+					}
+					doPublish(channel, replyTo, messageToSend);
+					return null;
+				}, this.recoveryCallback);
```

Comment:
Can we do here a wrapping to populate those attributes instead of `AtomicBoolean` logic in the `RetryCallback` impl above?
However no: we are going to have extra object for each message...

So, WDYT?